Dependency bumps

h-lame opened the pull request:
Upgrades:

* ruby to 2.3.7
* rspec to 3.7
* middleman and middleman-* to 4.x - this is the bulk of the effort as much changes from middleman 3 to 4 and lots needs shuffling about.

Best to check individual commits for the whys and wherefores.

h-lame committed:
Upgrade ruby to 2.3.7

To include the latest securtiy patches. Ideally we would be able to
upgrade to 2.5.x, but there is an open issue with rvm on Ubuntu 14.04
[1] which is preventing us from doing so.

[1] https://github.com/rvm/rvm/issues/4291

h-lame committed:
Upgrade rspec

h-lame committed:
Upgrade pry

h-lame committed:
Update middleman (and middleman-*) gems to latest

The site no longer works, we'll address those problems in later commits

h-lame committed:
Fix stylesheets to work in middleman 4

1. Add missing gems for sprockets. Middleman 4 no longer uses sprockets and
   compass by default, but we still use them so we have to add them back.
2. Add sprockets-globbing.  The version of middleman sprockets that works
   with middleman 4 no longer supports glob imports so we need to add a
   gem to get that functionality back. (e.g. `@import 'style/*'` instead
   of having to list each file in the folder).
3. We need the files to be named with their final file type, middleman no
   longer infers the filetype.  E.g. `application.scss` will become
   `application` so it needs to be called `application.css.scss` to become
   `application.css`.

h-lame committed:
Make sure all files have .html suffix

Seems that Middleman 3 would infer the final type of the file from context
or assume that everything was html unless told otherwise.  Middleman 4 no
longer does this so our files need to be `foo.html.<whatever>` not
`foo.<whatever>` if we want them to end up as html files.

h-lame committed:
Explicitly tell middleman-blog the source for articles

Seems that middleman 3's blogging engine would automatically pick up
the date from the title, but middleman 4's one doesn't seem to do so
and we have to explicitly define it.

h-lame committed:
Update middleman-blog-similar config for v2

With v2 we can no longer simply provide a custom algorithm, but we can add
a simple proc based tagger.  Our original algorithm would weight similar
tags and similar authors, so if we use the default tagger and add proc
that extracts the authors we should get similar results.

We can also remove our old algorithm because it's no longer used.

h-lame committed:
Fix after_build to work with middleman v4

Our after_build creates our redirects file that netlify will use to turn
requests into redirects.  Most of the methods we call on the `builder`
instance deal with files and these no longer live there.  In middleman v4
the `builder` is no longer a `thor` instance which is where those methods
came from.  Happily there is a `builder.thor` accessor we can use to get
these methods.

h-lame committed:
Make spinouts render the spinouts correctly again

To invoke Tilt directly we now need to provide a context (the current
middleman app) or helpers and such aren't available.

h-lame committed:
Fix custom author indexes

In middleman 4 we can't directly access data in config.rb as it's not
quite initialized yet.  Instead of an array we get a LazyCollectorStep.
In this case that means we don't detect any authors (as everyone ends up
as a NonEmployee) and so we don't generate the `/blog/author/person/`
pages for them.  If we switch to using `@app.data` instead then we get
the correct version of the data (e.g. an Array, not a lazy object) and
so our filter works again.

h-lame committed:
Add a note about the filter option on the author pages

It relies on our own custom filtering code that is not part of
middleman-blog.  While I was investigating this took me a while to
work out so a pointer would be useful.

tomsabin commented:
In https://github.com/unboxed/unboxed.co/pull/442/commits/07260ab6d1b7f92636175995940d3ad892ed8fdf (Fix stylesheets to work in middleman 4), I think the reset.scss file also needs `.css.scss` instead of just `.scss`?

```
source/assets/stylesheets/reset.css → source/assets/stylesheets/_reset.scss
```

h-lame commented:
We don't actually publish reset.css publicly so it's really a partial which is what the rename I've done makes clear, and as such that doesn't need to have a destination type extension.

Commit:
Remove additional / in path

This mistake didn't cause an error in the previous version, but it does
now.

Commit:
Add DefaultHelpers#partial monkey patch to render locals correctly

Since Middleman-4.1.0, the partial's local parameter is being modified.
This introduces breaking changes in a number of our partials, and so
we've fixed this in a monkey patch.

Commit:
Disable CircleCI database stage

ubxd commented:
Deploy preview for *unboxed-co-production* ready!

Built with commit c3334944ad862cb6a00e7d784246354820732d2b

https://deploy-preview-442--unboxed-co-production.netlify.com

h-lame commented on .ruby-version:
```diff
@@ -1 +1 @@
-ruby-2.5.1
+ruby-2.4.3
```
2.4.4 is latest in the 2.4.x branch and has a bunch of security vulns fixed so we should try that.  Probably useful to squash this down into my original 1st commit too.

tomsabin replied:
2.4.4 was also raising the same error as 2.5.1 unfortunately, so I've had to settle with 2.3.7 which includes the security vulnerability fixes.

<details>
<summary>Netlify build output</summary>

```
3:19:13 PM: Executing user command: middleman build --verbose
3:19:13 PM: /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- bundler/dep_proxy (LoadError)
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:888:in `block (2 levels) in expand_dependencies'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:887:in `each'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:887:in `block in expand_dependencies'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:874:in `each'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:874:in `expand_dependencies'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:806:in `converge_locked_specs'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:247:in `resolve'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:170:in `specs'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:237:in `specs_for'
3:19:13 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/definition.rb:226:in `requested_specs'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/runtime.rb:108:in `block in definition_method'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/runtime.rb:20:in `setup'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler.rb:107:in `setup'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/lib/bundler/setup.rb:20:in `<top (required)>'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
3:19:14 PM: 	from /opt/buildhome/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
3:19:14 PM: 	from /opt/build/cache/binstubs/middleman:27:in `<main>'
```

</details>
<br>

I've been speaking to Netlify's support, they are aware of the issues and are waiting for a fix for https://github.com/rvm/rvm/issues/4291.

h-lame commented on helpers/render_partial.rb:
```diff
@@ -0,0 +1,36 @@
+    # Bypass calling Middleman::CoreExtensions::DefaultHelpers#partial, and call
+    # Padrino::Helpers::RenderHelpers#partial instead.
+    method(__method__).super_method.super_method.call(template, options, &block)
+  end
```
I think, if possible, I'd rather we changed all our partials to work around this issue if that's even possible.  If we have to rely on the monkey patch can we put some kind of timebomb in here to warn us to check if we still need it when we upgrade.  

Is it possible to raise an issue against middleman so that they might fix this upstream?

tomsabin replied:
> Is it possible to raise an issue against middleman so that they might fix this upstream?

Yes, as the fix is quite simple, I'm planning to submit a PR to Middleman to fix this, as it's not the first time it's been an issue: https://github.com/middleman/middleman/issues/1981

tomsabin replied:
> can we put some kind of timebomb in here to warn us to check if we still need it when we upgrade

I'm not too sure what this'll look like.. do you mean something like raise an error if the middleman version has been changed?

h-lame replied:
Yeah something like that.  Here's one I did aaaaages ago: https://github.com/alphagov/whitehall/commit/7280ece8c6cfdf1914c85443cf56370493158d2b

Commit:
Raise runtime error if Middleman is updated again

We've added a monkey patch to fix Middleman's DefaultHelpers#partial
method. However, we're expecting that this is resolved in a later
version and that eventually this helper can be removed.

h-lame commented:
If we rebase again we can drop 338ce237dc55995fab743af4fb291e47654cafe3 because there's a version already on master.

tomsabin commented:
It seems CircleCI treats the existing circle.yml database configuration differently for this branch than what's on production.

Without https://github.com/unboxed/unboxed.co/commit/338ce237dc55995fab743af4fb291e47654cafe3, the empty database override doesn't seem to take effect, causing the build to fail at the database setup section. I've added it back in, and it should now be clearer with the failing/passing builds.

Commit:
Add link in monkey patch version-bomb to Middleman PR
